Test/add category test cases

monaayman123 opened the pull request:
# Pull Request Template

## Description
This PR adds a complete set of unit tests across all layers Remote, Repository, Use Cases, and ViewModels  for both Movies and TV Shows category details.
---

## Changes Made
List the changes introduced in this pull request:
1️⃣ Remote Layer Tests
Added tests for Remote Data Sources to verify:

Successful API calls returning valid data.

Handling of errors such as network failures or invalid responses.

DTO mapping correctness.

Functions tested:

getMoviesByGenreIdRemote

getTvShowsByGenreIdRemote

2️⃣ Repository Layer Tests
Used MockWebServer to mock API responses.

Verified that the Repository:

Calls the Remote Data Source with the correct parameters.

Maps DTO → Entity correctly.

Throws AflamiException for error scenarios.

Functions tested:

getMoviesByGenreId

getTvShowsByGenreId

3️⃣ Use Case Tests
Tested GetMoviesByGenreUseCase and GetTvShowsByGenreUseCase:

Ensured they delegate correctly to the Repository.

Verified correct results or exception propagation.

Functions tested:

invoke(selectedGenre, page) for:

GetMoviesByGenreUseCase

GetTvShowsByGenreUseCase

4️⃣ ViewModel Tests
CategoriesMoviesDetailsViewModelTest
Mocked GetMoviesByGenreUseCase and CategoriesMovieDetailsArgs.

Verified UI state and effect flows.

Functions tested:

onClickBack → emits NavigateBack.

onClickMovieCard(movieId) → emits NavigateToMovieDetails(movieId).

onClickGenre (same genre) → no reload triggered.

onClickGenre (different genre) → updates state & triggers reload.

onClickRetryRequest → triggers reload.

onPagingLoadStateChanged:

LoadState.Loading → isLoading = true.

LoadState.NotLoading → isLoading = false.

LoadState.Error → updates errorUiState & sets isLoading = false.

CategoriesTvShowsDetailsViewModelTest
Same structure as Movies ViewModel tests but for TV Shows.

Functions tested:

onClickBack → emits NavigateBack.

onClickTvShowCard(tvShowId) → emits NavigateToTvShowDetails(tvShowId).

onClickGenre (same genre) → no reload triggered.

onClickGenre (different genre) → updates state & triggers reload.

onClickRetryRequest → triggers reload.

onPagingLoadStateChanged:

LoadState.Loading → isLoading = true.

LoadState.NotLoading → isLoading = false.

LoadState.Error → updates errorUiState & sets isLoading = false.


## Checklist
Please ensure the following tasks are completed:

- [ ] My code follows the code style of this project
- [ ] Changes have been tested manually and verified.
- [ ] PR includes at most one single feature.

---

## Additional Comments
Add any additional information or context about the pull request here.

monaayman123 committed:
test: GetTvShowsByGenreUseCaseTest.kt

monaayman123 committed:
test: GetMoviesByGenreUseCaseTest.kt

monaayman123 committed:
test: refactor data type of MovieId to e long

monaayman123 committed:
test: add test case for getMoviesByGenreId

monaayman123 committed:
test: add test case getTvShowByGenreId

monaayman123 committed:
test: update id to be 28L

monaayman123 committed:
test: fix broken test cases

monaayman123 committed:
test: fix broken test cases

monaayman123 committed:
test: add test case for getMoviesByGenre

monaayman123 committed:
test: add test case for getMoviesByGenre

monaayman123 committed:
test: add test case for getTvShowByGenre

monaayman123 committed:
test: fix broken test cases

monaayman123 committed:
test: add CategoriesTvShowsDetailsViewModelTest.kt

monaayman123 committed:
refactor: rename NavigateToCategoriesDetailsScreen to be NavigateToCategoriesMoviesDetailsScreen

monaayman123 committed:
refactor: add CategoriesViewModelTest.kt

monaayman123 committed:
refactor: add CategoriesMoviesDetailsViewModelTest.kt

monaayman123 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into test/add-category-test-cases

saraashraf11 commented on remoteDatasource/src/test/java/com/amsterdam/remotedatasource/datasource/MovieRemoteDataSourceImplTest.kt:
```diff
@@ -461,6 +461,26 @@ class MovieRemoteDataSourceImplTest {
+
+    @Test
+    fun `getMoviesByGenreId should return movies for given genre ID when successful `() = runTest {
+        // Given
```
Delete this comment in all files

saraashraf11 commented on remoteDatasource/src/test/java/com/amsterdam/remotedatasource/datasource/TvRemoteDataSourceImplTest.kt:
```diff
@@ -32,7 +32,6 @@ class TvRemoteDataSourceImplTest {
     @Test
     fun `getPopularTvShows should return a list of TV shows when successful`() = runTest {
-        // Given
         val expectedResponse = RemoteTvShowResponse(
```
According to the code review from Falah and Asia's words, we need to separate the entity in a separate file or even at the end of this file and use it here.

ThoriaHamdy99 replied:
we agreed they will be at the end of the file

saraashraf11 commented on remoteDatasource/src/test/java/com/amsterdam/remotedatasource/datasource/TvRemoteDataSourceImplTest.kt:
```diff
@@ -43,28 +42,23 @@ class TvRemoteDataSourceImplTest {
     @Test
     fun `getPopularTvShows should return an empty list when API response is empty`() = runTest {
-        // Given
         val expectedResponse = RemoteTvShowResponse(
```
You already need it in more than one place, so it is better to separate it.

ThoriaHamdy99 commented on domain/src/test/java/com/amsterdam/domain/useCase/details/GetMoviesByGenreUseCaseTest.kt:
```diff
@@ -0,0 +1,38 @@
+
+    @BeforeEach
+    fun setUp() {
+        movieRepository = mockk()
```
no need for the repo and use case to be initialize inside the setup. It can be initialize by lazy

ThoriaHamdy99 commented on domain/src/test/java/com/amsterdam/domain/useCase/details/GetMoviesByGenreUseCaseTest.kt:
```diff
@@ -0,0 +1,38 @@
+
+    @Test
+    fun `invoke should call getMoviesByGenre from movieRepository`() = runTest {
+        val selectedGenre = MovieGenre.ACTION
```
Please move these variables outside the test function and at the end of the file

ThoriaHamdy99 replied:
You can check the dao tests to see what I mean

ThoriaHamdy99 commented on domain/src/test/java/com/amsterdam/domain/useCase/details/GetMoviesByGenreUseCaseTest.kt:
```diff
@@ -0,0 +1,38 @@
+        val result = getMoviesByGenreUseCase(selectedGenre, page)
+
+        assertThat(result).isEqualTo(movies)
+        coVerify { movieRepository.getMoviesByGenre(selectedGenre, page) }
```
The verification need to be moved in another test case

monaayman123 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into test/add-category-test-cases

# Conflicts:
#	repository/src/test/java/com/amsterdam/repository/repository/TvShowRepositoryImplTest.kt
#	repository/src/test/java/com/amsterdam/repository/repository/UserListRepositoryImplTest.kt
#	ui/src/main/java/com/amsterdam/ui/screens/categories/CategoriesScreen.kt

monaayman123 committed:
test: move variable definations outside function and initialize repo and use case by lazy

monaayman123 committed:
test: fix broken test

monaayman123 committed:
test: move variable definations outside function and initialize repo and use case by lazy

monaayman123 committed:
test: move variable definations outside function and initialize repo and use case by lazy

monaayman123 committed:
test: move variable definations outside function and initialize repo and use case by lazy

monaayman123 committed:
test: move variable definations outside function and initialize repo and use case by lazy

ThoriaHamdy99 commented on domain/src/test/java/com/amsterdam/domain/useCase/details/GetMoviesByGenreUseCaseTest.kt:
```diff
@@ -0,0 +1,40 @@
+
+class GetMoviesByGenreUseCaseTest {
+    private val movieRepository: MovieRepository = mockk()
+    val getMoviesByGenreUseCase by lazy {
```
make it private

ThoriaHamdy99 commented on domain/src/test/java/com/amsterdam/domain/useCase/details/GetMoviesByGenreUseCaseTest.kt:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @Test
+    fun `invoke should call getMoviesByGenre from movieRepository`() = runTest {
```
the naming of the function has no mean for what it does.

the function here not verifying it make an assertion so the function here try to ensure about the returned data 
please handle its naming

ThoriaHamdy99 commented on domain/src/test/java/com/amsterdam/domain/useCase/details/GetMoviesByGenreUseCaseTest.kt:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @Test
+    fun `should call getMoviesByGenre from movieRepository`() = runTest {
```
make the test function name starts with the function that you test

ThoriaHamdy99 commented on domain/src/test/java/com/amsterdam/domain/useCase/details/GetMoviesByGenreUseCaseTest.kt:
```diff
@@ -0,0 +1,40 @@
+
+    @Test
+    fun `should call getMoviesByGenre from movieRepository`() = runTest {
+        coEvery { movieRepository.getMoviesByGenre(selectedGenre, page) } returns mockk()
```
try to make anew line after the given part and when part. to separate them for more redability

ThoriaHamdy99 commented on domain/src/test/java/com/amsterdam/domain/useCase/details/GetMoviesByGenreUseCaseTest.kt:
```diff
@@ -0,0 +1,40 @@
+
+    @Test
+    fun `invoke should call getMoviesByGenre from movieRepository`() = runTest {
+
```
No need for this line

ThoriaHamdy99 commented on domain/src/test/java/com/amsterdam/domain/useCase/details/GetMoviesByGenreUseCaseTest.kt:
```diff
@@ -0,0 +1,40 @@
+        coVerify(exactly = 1) { movieRepository.getMoviesByGenre(selectedGenre, page) }
+    }
+
+    val selectedGenre = MovieGenre.ACTION
```
make these variables private

ThoriaHamdy99 commented on domain/src/test/java/com/amsterdam/domain/useCase/details/GetTvShowsByGenreUseCaseTest.kt:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    @Test
+    fun `invoke should call getTvShowsByGenre from tvShowRepository`() = runTest {
```
try to handle the function name to tell ue what it does

ThoriaHamdy99 commented on domain/src/test/java/com/amsterdam/domain/useCase/details/GetTvShowsByGenreUseCaseTest.kt:
```diff
@@ -0,0 +1,36 @@
+        coVerify { tvShowRepository.getTvShowsByGenre(selectedGenre, page) }
+
+    }
+    val selectedGenre = TvShowGenre.COMEDY
```
private variables

ThoriaHamdy99 commented on remoteDatasource/src/test/java/com/amsterdam/remotedatasource/datasource/TvRemoteDataSourceImplTest.kt:
```diff
@@ -303,74 +226,79 @@ class TvRemoteDataSourceImplTest {
+    @Test
+    fun `getTvShowByGenreIds should return TV shows for given genre IDs when successful`() =
+        runTest {
+            val genreIds = listOf(28L)
```
put this variables as private variables at the end of the file

ThoriaHamdy99 commented on remoteDatasource/src/test/java/com/amsterdam/remotedatasource/datasource/TvRemoteDataSourceImplTest.kt:
```diff
@@ -303,74 +226,79 @@ class TvRemoteDataSourceImplTest {
+
+        }
+
+    val keyword = "Game of Thrones"
```
also here for these variables

ThoriaHamdy99 reviewed (changes requested):
Please, Move all variables outside the test functions let it be private variables.
Handle the function names to start with function then should part then when part.
Try to make a new line after the given part and after when part. to separate them for more readability without any comments

monaayman123 committed:
test: move variable definations outside function and make them private and initialize repo and use case by lazy

monaayman123 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into test/add-category-test-cases

# Conflicts:
#	remoteDatasource/src/test/java/com/amsterdam/remotedatasource/datasource/MovieRemoteDataSourceImplTest.kt
#	remoteDatasource/src/test/java/com/amsterdam/remotedatasource/datasource/TvRemoteDataSourceImplTest.kt
#	remoteDatasource/src/test/java/com/amsterdam/remotedatasource/datasource/UserListRemoteDataSourceImplTest.kt
#	remoteDatasource/src/test/java/com/amsterdam/remotedatasource/util/CreateRemoteFakeData.kt
#	repository/src/test/java/com/amsterdam/repository/repository/MovieRepositoryImplTest.kt
#	repository/src/test/java/com/amsterdam/repository/repository/TvShowRepositoryImplTest.kt
#	repository/src/test/java/com/amsterdam/repository/repository/UserListRepositoryImplTest.kt

monaayman123 committed:
test: resolve conflicts

ThoriaHamdy99 commented on viewModel/src/test/java/com/amsterdam/viewmodel/categories/CategoriesMoviesDetailsViewModelTest.kt:
```diff
@@ -0,0 +1,123 @@
+
+    @Test
+    fun `onPagingLoadStateChanged with NotLoading should set isLoading false`() = runTest {
+        val loadStates = CombinedLoadStates(
```
make this outside the test function

ThoriaHamdy99 commented on viewModel/src/test/java/com/amsterdam/viewmodel/categories/CategoriesMoviesDetailsViewModelTest.kt:
```diff
@@ -0,0 +1,123 @@
+
+    @Test
+    fun `onPagingLoadStateChanged with Loading should set isLoading true`() = runTest {
+        val loadStates = CombinedLoadStates(
```
make this outside the test function

ThoriaHamdy99 commented on viewModel/src/test/java/com/amsterdam/viewmodel/categories/CategoriesTvShowsDetailsViewModelTest.kt:
```diff
@@ -0,0 +1,93 @@
+
+    @Test
+    fun `onClickGenre with different genre should update selectedGenre`() = runTest {
+        val newGenre = TvShowGenre.COMEDY
```
make this outside the test function

monaayman123 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into test/add-category-test-cases

# Conflicts:
#	domain/src/test/java/com/amsterdam/domain/useCase/details/GetMoviesByGenreUseCaseTest.kt
#	domain/src/test/java/com/amsterdam/domain/useCase/details/GetTvShowsByGenreUseCaseTest.kt
#	domain/src/test/java/com/amsterdam/domain/useCase/list/AddMovieToListUseCaseTest.kt
#	domain/src/test/java/com/amsterdam/domain/useCase/profile/GetUserPointsUseCaseTest.kt
#	remoteDatasource/src/test/java/com/amsterdam/remotedatasource/datasource/MovieRemoteDataSourceImplTest.kt
#	remoteDatasource/src/test/java/com/amsterdam/remotedatasource/datasource/ProfileRemoteDataSourceImplTest.kt
#	remoteDatasource/src/test/java/com/amsterdam/remotedatasource/datasource/TvRemoteDataSourceImplTest.kt
#	remoteDatasource/src/test/java/com/amsterdam/remotedatasource/datasource/UserListRemoteDataSourceImplTest.kt
#	remoteDatasource/src/test/java/com/amsterdam/remotedatasource/util/CreateRemoteFakeData.kt
#	repository/src/test/java/com/amsterdam/repository/repository/UserListRepositoryImplTest.kt

monaayman123 committed:
test: resolve conflicts

monaayman123 committed:
test: move variables outside functions

monaayman123 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into test/add-category-test-cases

monaayman123 committed:
test: Corrected multiple failing and flaky ViewModel unit tests.

- Implemented a MainDispatcherExtension to control Dispatchers.Main and fix coroutine timing issues that caused Turbine timeouts.
 - Fixed initialization order bugs that led to IllegalArgumentExceptions.
 - Added `advanceUntilIdle()` where necessary to wait for asynchronous state updates before making assertions.

monaayman123 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into test/add-category-test-cases

ThoriaHamdy99 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into test/add-category-test-cases

# Conflicts:
#	remoteDatasource/src/test/java/com/amsterdam/remotedatasource/datasource/MovieRemoteDataSourceImplTest.kt

ThoriaHamdy99 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into test/add-category-test-cases

ThoriaHamdy99 reviewed (changes requested):
Please change the naming of the function to match the approach that we follow

ThoriaHamdy99 commented on domain/src/test/java/com/amsterdam/domain/useCase/details/GetMoviesByGenreUseCaseTest.kt:
```diff
@@ -18,7 +18,7 @@ class GetMoviesByGenreUseCaseTest {
 
     @Test
-    fun `should call getMoviesByGenre from movieRepository when invoked`() =
+    fun `when getMoviesByGenreUseCase is called, should call getMoviesByGenre from movieRepository`() =
```
@monaayman123 
Here the naming should be starts with the function name that you test then should part then when part

Like --> getMoviesByGenre should call getMoviesByGenre from movieRepository when getMoviesByGenreUseCase is called

monaayman123 replied:
done

monaayman123 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into test/add-category-test-cases

# Conflicts:
#	remoteDatasource/src/test/java/com/amsterdam/remotedatasource/datasource/MovieRemoteDataSourceImplTest.kt

ThoriaHamdy99 reviewed (changes requested):
Please run the app after pull from develop as there is a problem in your branch, I think conflict after merging you need to fix it.

monaayman123 committed:
refactor: rename function

monaayman123 committed:
Merge remote-tracking branch 'origin/test/add-category-test-cases' into test/add-category-test-cases

# Conflicts:
#	remoteDatasource/src/test/java/com/amsterdam/remotedatasource/datasource/MovieRemoteDataSourceImplTest.kt

monaayman123 committed:
refactor: remove unneeded files

monaayman123 committed:
refactor: add missing braket

monaayman123 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into test/add-category-test-cases
